propertyestimator/layers/reweighting.py

```diff
@@ -42,24 +90,188 @@ def schedule_calculation(calculation_backend, storage_backend, layer_directory,
+            reduced_potentials.append(potential_energy / unit.kilojoule_per_mole)
+
+        if thermodynamic_state.pressure is not None:
+
+            volumes = existing_data.statistics_data.get_observable(ObservableType.Volume)
+
+            for index, volume in enumerate(volumes):
+
+                reduced_potentials[index] += (volume * thermodynamic_state.pressure *
+                                              unit.AVOGADRO_CONSTANT_NA) / unit.kilojoule_per_mole
+
+        beta = 1.0 / (thermodynamic_state.temperature * unit.MOLAR_GAS_CONSTANT_R)
+
+        for index, reduced_potential in enumerate(reduced_potentials):
+            reduced_potentials[index] = (reduced_potential * unit.kilojoule_per_mole) * beta
+
+        return np.array(reduced_potentials)
+
+    @staticmethod
+    def resample_data(substance, trajectory_to_resample,
+                      force_field, available_resources):
+        """Resample the frames of a trajectory using a different set
+        of force field parameters than were used when generating the
+        trajectory.
+
+        Parameters
+        ----------
+        substance: Mixture
+            The substance being resampled.
+        trajectory_to_resample: :obj:`mdtraj.Trajectory`
+            The trajectory to resample.
+        thermodynamic_state: :obj:`propertyestimator.thermodynamics.ThermodynamicState`
+            The thermodynamic state to resample the statistics at.
+        force_field: :obj:`openforcefield.typing.engines.smirnoff.ForceField`
+            The force field to use when calculating the energy of the frames of
+            the trajectory.
+
+        Returns
+        -------
+        :obj:`numpy.ndarray`
+            The resampled energies.
         """
 
-        # A return value indicates that the reweighting layer did not
-        # have access to enough information to accurately estimate the property.
-        return None
+        molecules = [create_molecule_from_smiles(component.smiles) for component in substance.components]
+
+        from openforcefield.typing.engines import smirnoff
+        from simtk.openmm import Platform, Context, VerletIntegrator
+
+        topology = trajectory_to_resample.topology.to_openmm
+
+        system = force_field.createSystem(topology,
+                                          molecules,
+                                          nonbondedMethod=smirnoff.PME,
+                                          chargeMethod='OECharges_AM1BCCSym')
+
+        if system is None:
+
+            return PropertyEstimatorException(directory='',
+                                              message='Failed to create a system from the'
+                                                      'provided topology and molecules')
+
+        context = None
+        integrator = VerletIntegrator(0.002 * unit.picoseconds)
+
+        if available_resources.number_of_gpus > 0:
+
+            gpu_platform = Platform.getPlatformByName('CUDA')
```

Comment:
We should never hardcode the `CUDA` platform, since it will lead to exceptions if the wrong CUDA toolkit version is installed or no GPUs are installed. If we don't specify a platform, we can instead fall back gracefully to `OpenCL` or `CPU` platforms. It's a bit more complicated to create a platform and make sure we're using both deterministic PME and mixed precision context properties, since these aren't the same on all platforms, but @andrrizzi has essentially hidden all of this detail for us under the hood of his [`ContextCache`](https://openmmtools.readthedocs.io/en/0.17.0/api/generated/openmmtools.cache.ContextCache.html#openmmtools.cache.ContextCache) and [`ThermodynamicState`](https://openmmtools.readthedocs.io/en/0.17.0/api/generated/openmmtools.states.ThermodynamicState.html#openmmtools.states.ThermodynamicState) objects, which we should use whenever possible.

Reply:
In this case I'd argue we definitely do want the exception to be raised. When you set up the `PropertyEstimatorServer` you specify which compute resources (number of cores, gpus etc) you want each worker to have, with the expectation that the worker will make full use of these resources when possible. We don't want openmm (and thus the estimator) to silently fall back to using threads because of a bad cuda installation, or because there are no GPUs as this would be contrary to the expected user experience, and we should inform the user that what they are asking for could not be completed so they can remedy it.

Reply:
I'll definitely add a flag to the resources object however to let the user choose between CUDA and OpenCL.

Reply:
You're right about not wanting to fall back to CPU threads, but the CUDA version match requirements are so finicky that I *do* think we want to be able to fall back to OpenCL if CUDA is unavailable. 

We also *definitely* want to make sure the right platform parameters get set to ensure we have deterministic forces and mixed precision.

And in general, instead of hardcoding a lot of this general functionality, it would be great to encapsulate it in general management helper classes like [`ContextCache`](https://openmmtools.readthedocs.io/en/0.17.0/api/generated/openmmtools.cache.ContextCache.html#openmmtools.cache.ContextCache).

Reply:
I think what you propose is a good way forward! I'm working on refactoring the reweighting code to be more modular and to take advantage of the protocol workflows and the openmmtools utilities, so will make sure to include these changes also.